File: Smartpoint.py
from datetime import datetime
import os
import threading

#LogFileManager manager
class LogFileManager:
    def __init__(self):
        self._files = {}
        self._Lock = threading.Lock()

    #로그 파일 열기
    def openLogFile(self, filename: str):
        with self._Lock:
            if filename not in self._files:
                try:
                    f = open(filename, 'a', encoding='utf-8')
                    self._files[filename] = f
                except Exception as e:
                    raise RuntimeError(f"파일 열기 실패: {filename} ({e})")
                
    #로그 파일 쓰기
    def writeLog(self, filename: str, message: str):
        Dtime = datetime.now().strftime("[%y-%m-%d %h:%m:%s]")
        with self._Lock:
            if filename not in self._files:
                raise RuntimeError(f"파일이 열려있지 않음: {filename}")
            try:
                self._files[filename].write(f"{Dtime} {message}\n")
                self._files[filename].flush()
            except Exception as e:
                raise RuntimeError(f"로그 쓰기 실패: {filename} ({e})")
            
    #로그 파일 읽기
    def readLogs(self, filename: str):
        pass


    #로그 파일 닫기
    def closeLogFile(self, filename: str):
        pass

File: Smartpoint.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <map>
#include <vector>
#include <memory>
#include <ctime>
#include <sstream>
#include <iomanip>
#include <stdexcept>

class LogFileManager {
private:
    // 파일 이름별 ofstream 스마트 포인터 관리
    std::map<std::string, std::unique_ptr<std::ofstream>> fileMap;

    // 타임스탬프 생성 함수
    std::string currentTimestamp() {
        std::ostringstream oss;
        auto t = std::time(nullptr);
        auto tm = *std::localtime(&t);
        oss << "[" << std::put_time(&tm, "%Y-%m-%d %H:%M:%S") << "]";
        return oss.str();
    }

public:
    LogFileManager() = default;
    ~LogFileManager() {
        // 모든 파일을 안전하게 닫기
        for (auto& pair : fileMap) {
            if (pair.second && pair.second->is_open()) {
                pair.second->close();
            }
        }
    }

    // 복사/이동 금지
    LogFileManager(const LogFileManager&) = delete;
    LogFileManager& operator=(const LogFileManager&) = delete;

    // 로그 파일 열기
    void openLogFile(const std::string& filename) {
        if (fileMap.count(filename) == 0) {
            auto ofs = std::make_unique<std::ofstream>(filename, std::ios::app);
            if (!ofs->is_open()) {
                throw std::runtime_error("파일 열기 실패: " + filename);
            }
            fileMap[filename] = std::move(ofs);
        }
    }

    // 로그 파일에 메시지 쓰기 (타임스탬프 포함)
    void writeLog(const std::string& filename, const std::string& message) {
        auto it = fileMap.find(filename);
        if (it == fileMap.end() || !it->second || !it->second->is_open()) {
            throw std::runtime_error("파일이 열려있지 않습니다: " + filename);
        }
        *(it->second) << currentTimestamp() << " " << message << std::endl;
    }

    // 로그 파일에서 전체 라인들을 vector로 반환
    std::vector<std::string> readLogs(const std::string& filename) {
        std::vector<std::string> result;
        std::ifstream ifs(filename);
        if (!ifs.is_open()) {
            throw std::runtime_error("파일 읽기 실패: " + filename);
        }
        std::string line;
        while (std::getline(ifs, line)) {
            if (!line.empty()) result.push_back(line);
        }
        return result;
    }

    // 로그 파일 닫기
    void closeLogFile(const std::string& filename) {
        auto it = fileMap.find(filename);
        if (it != fileMap.end() && it->second && it->second->is_open()) {
            it->second->close();
            fileMap.erase(it);
        }
    }
};

int main() {
    LogFileManager manager;
    manager.openLogFile("error.log");
    manager.openLogFile("debug.log");
    manager.openLogFile("info.log");

    manager.writeLog("error.log", "Database connection failed");
    manager.writeLog("debug.log", "User login attempt");
    manager.writeLog("info.log", "Server started successfully");

    std::vector<std::string> errorLogs = manager.readLogs("error.log");
    
    for (const auto& line : errorLogs) {
        
    }

    manager.closeLogFile("error.log");
    manager.closeLogFile("debug.log");
    manager.closeLogFile("info.log");
    return 0;
}

